### DISCUSSION TITLE
Potential to port v360 functionality from ffmpeg?

### DISCUSSION THREAD

**Discussion by richardtallent** [libvips/libvips#2573] [Ideas]
Hi,

I'm already using libvips (via the node sharp package) to automate various bits of my image-handling workflow. It works beautifully, saves me a TON of time in PhotoShop, especially creating web-optimized images.

Now I'm working on an art series of photos taken with a circular fisheye lens, and I need to convert them to an equirectangular projection. There doesn't seem to be functionality built into libvips, but after searching high and low, I found that the `v360` filter in ffmpeg does a good job of this and many similar projection transformations.

I suspect I'm not the only photographer out here needing this, especially as people get more interested in AR/VR applications.

What would be the potential for porting this functionality from ffmpeg into libvips? Both are LGPL-licensed, and both are C-based (though ffmpeg does have a few x86 assembly routines for performance).

https://github.com/FFmpeg/FFmpeg/blob/a7dfa6b446e8ced9f25efd0597e6073e6b22a94b/libavfilter/vf_v360.c

Are there other areas where these projects could coordinate?

**Comment by jcupitt** [comment]
Hi @richardtallent,

libvips has something like this via `mapim`. There are some examples here:

http://libvips.blogspot.com/2015/11/fancy-transforms.html

Would that do what you need?

**Comment by richardtallent** [comment]
Interesting!

The approach is different, but it could be the basis for something similar. However, it would leave the implementation of creating the index images for various transforms to the user. This is a big advantage of that ffmpeg filter--all that math is done already, you just throw `v360=input=fisheye:ih_fov=180:iv_fov=180:e` in the command line and Bob's your uncle.

I *think* `mapim` also only does bilinear interpolation, which could lead to a suboptimal result. There's a fair amount of interpolation, so, especially with all of the additional re-mapping needed to display on a VR display (and possibly as one side of a stereo pair), being able to choose the algorithm might make a real difference in output quality.

**Reply by jcupitt** [reply]
I've not checked, but the equations are pretty simple I think. It'd just be a line or two of python.

You can set the mapim interpolator, though the interpolator classes always use a fixed stencil, so you could get aliasing on very large downsamples.
